localization/ndt_scan_matcher/include/ndt_scan_matcher/map_update_module.hpp

```diff
@@ -44,36 +45,46 @@ class MapUpdateModule
   using PointTarget = pcl::PointXYZ;
   using NormalDistributionsTransform =
     pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>;
+  using NdtPtrType = std::shared_ptr<NormalDistributionsTransform>;
 
 public:
   MapUpdateModule(
     rclcpp::Node * node, std::mutex * ndt_ptr_mutex,
-    std::shared_ptr<NormalDistributionsTransform> ndt_ptr,
+    std::shared_ptr<NormalDistributionsTransform> & ndt_ptr,
     HyperParameters::DynamicMapLoading param);
 
 private:
   friend class NDTScanMatcher;
 
+  // Update the specified NDT
   void update_ndt(
+    NdtPtrType & ndt,
     const std::vector<autoware_map_msgs::msg::PointCloudMapCellWithID> & maps_to_add,
     const std::vector<std::string> & map_ids_to_remove);
   void update_map(const geometry_msgs::msg::Point & position);
-  [[nodiscard]] bool should_update_map(const geometry_msgs::msg::Point & position) const;
+  [[nodiscard]] bool should_update_map(const geometry_msgs::msg::Point & position);
   void publish_partial_pcd_map();
 
+  // Pre-fetch NDT map in a child thread
+  void prefetch_map(const geometry_msgs::msg::Point & position, NdtPtrType & ndt);
+
   rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr loaded_pcd_pub_;
 
   rclcpp::Client<autoware_map_msgs::srv::GetDifferentialPointCloudMap>::SharedPtr
     pcd_loader_client_;
 
-  std::shared_ptr<NormalDistributionsTransform> ndt_ptr_;
+  NdtPtrType & ndt_ptr_;
   std::mutex * ndt_ptr_mutex_;
   rclcpp::Logger logger_;
   rclcpp::Clock::SharedPtr clock_;
 
   std::optional<geometry_msgs::msg::Point> last_update_position_ = std::nullopt;
 
   HyperParameters::DynamicMapLoading param_;
+
+  // Indicate if there is a prefetch thread waiting for being collected
+  NdtPtrType secondary_ndt_ptr_;
+  bool need_rebuild_;
```

Comment:
After discussing it with Ando-san, we came to the conclusion that `need_rebuild_` may not be necessary for this pull request. What do you think?

Even if this module needs to rebuild `ndt_ptr_` itself, I thought it is possible to update the `secondary_ndt_ptr_` and copying it to `ndt_ptr_`.
There will be waste copying, but I think it will eliminate conditional branches and make the code simpler.

Please let me know if you have any questions.

Reply:
@SakodaShintaro There are two cases here:

1. The current position is close to the previously loaded position. In this case, aligning the current scan to the old `ndt_ptr_` still works, because the scan and the previously loaded map overlap to some extent. Hence, we can update the `secondary_ndt_ptr_` while aligning the current scan with the old `ndt_ptr_` just fine.
2. The current position is very far from the previously loaded position. In this case, aligning the current scan to the old ndt_ptr_ produces the incorrect result, because the scan the previously loaded map does not overlap at all. If we does not lock ndt_ptr_ in this case, the system will try to align the incoming scans (until the secondary_ndt_ptr_ is 100% updated) with the old ndt_ptr_, thus giving incorrect results. Therefore, in this case, I think the ndt_ptr_ must be locked and rebuilt before any scan alignment is called.

Reply:
I also understand the need for `need_rebuild_`.
The code is OK as is.